### PR TITLE
QVAC-24335 infra: install mobile prebuilds from a run id and use prebuild to run mobile tests instead of Publish GPR

### PR DESCRIPTION
**Pull request by tobi-legan**
## 🎯 What problem does this PR solve?

- Testing an unmerged **native** change on a device took four manual steps after mobile became dispatch-only (QVAC-23665): push a `tmp-*` branch, dispatch On Merge Trigger, wait for the publish, then hand-assemble `@tetherto/<addon>-mono@<version>-tmp.runid-<id>` as `package`.

- Skipping any of them silently tested the **published release** instead. QVAC-24253 ran mobile on 5 addons and went green with `package` empty, so every run installed the published tarball rather than the ~300 lines of new C++ under review. Nobody noticed, because the input help said empty meant "this branch's prebuild artifact".

- `@qvac/asr-ggml` and `@qvac/audiogen-ggml` publish **no mobile prebuilds at all**, so for those two an empty input could never work — it fails with `No prebuilds directory found in package`.

## 📝 How does it solve it?

- New `prebuild_run_id` dispatch input installs the `prebuilds-<pkg>` artifact a named run already built (falling back to the legacy `prebuilds` name). Point it at the on-pr run that built your prebuilds — no tmp branch, no publish, no package name.

- Highest-precedence prebuild source. Every other source is gated on an empty run id, and `prebuilds/` is cleared before extraction, so nothing can shadow the resolved run.

- **Fails closed.** A run id that is wrong, unreadable, unfinished or expired fails the run with the reason instead of sliding back to `@latest`. Setting both `prebuild_run_id` and `package` is refused rather than silently resolved.

- The download selects the artifact by **id**, not name: a re-run leaves the earlier attempt's artifacts under the same run id, so a run can hold two live `prebuilds-<pkg>` rows.

- Provenance names the run, artifact, head SHA, branch and the repository the binaries were built from. A fork-built run **warns rather than failing** — this repo is fork-first (`docs/gitflow.md`), and a fork's prebuilds only exist once the merge/release team approved `fork-ci` on that run.

- `audiogen-ggml` pins its composite actions to the default branch, so it ignores an unknown input with only a warning and would silently install `@latest`. The setup action now exposes `prebuild-source-run-id` and that workflow asserts it matches the request, failing with instructions instead. A test requires the same assertion from any workflow pinning the composite to the default branch.

- Node, not `gh`: these jobs run on self-hosted runners where the CLI is not guaranteed. The resolver sits beside the action so the callers' existing sparse checkout covers it.

- Adds the `qv-mobile-test-dispatch` skill (`.agents/skills/`, discovered directly by Codex and Cursor, linked into `.claude/skills` by `scripts/agent-setup.sh`) and a quick-start in `docs/ci/MOBILE-ON-DEMAND.md`, both covering how to find a run id, pick a valid test filter, and **where the Android and iOS logs are, including the native C++ output**.

**Not applicable to two addons, by design:** `decoder-audio` has no native prebuild of its own (it rides `bare-ffmpeg` from npm) and `inference-addon-cpp` compiles its own prebuilds in-run from the dispatched ref. A test pins both reasons so a third exemption has to be a decision.

**Interface change is additive:** `prebuild-run-id` is optional and defaults to `""`, and `prebuild-source-run-id` is a new output. Every existing caller is unchanged — verified by control runs that still resolve `@qvac/<addon>@latest` from npm. Not tagged `[bc]`.

### Drive-by fixes found while testing

- **`vla-ggml` prestage drift** — the on-device drift oracle was built from `MODEL_SHARDS` (the runners that *have* a model) instead of `test-groups.json` (the runners that *exist*), so any model-less runner failed closed with `[prestage] FATAL: matched no known runner`. `runEsmNamedExportsTest` is scheduled on both platforms and the mobile-deferred `runPi05Test` would have failed the same way.
- **`model-fit` fixture path** — `fit-stub.test.js` wrote fixtures to a second derivation of "beside the model" (`path.resolve(__dirname, '../model')`), which on device points inside the read-only app bundle: `ENOENT: ... stat "/app.bundle"`. Now derived from the model's own path.
- **Mobile log diagnostics** — `flushBareLog` fed Appium's error *object* to `Buffer.from`, so every failure surfaced as `The first argument must be of type string…`, masking the real reason. Now type-checked and reported; the Android branch states plainly that `bare_console.log` does not exist there and points at `logcat_full.txt` under the `bare` tag.

### Follow-up: the PR-head TypeScript gate

Six of this PR's own checks were red for a reason worth fixing rather than
working around, and the fix landed here.

`<pkg>-pr-head-ts-checks` is published by `on-pr-ts-nx` only when nx finds the
package's TypeScript affected. A PR that touches a package's WORKFLOW but not
its code leaves that producer job skipped — and the awaiters treated a skip as a
failure, so such a PR was unmergeable. Fixing the conclusion was only half of
it: the callers poll for the job-level name a reusable-workflow call produces
(`llm-pr-head-ts-checks / ts-checks`), but when the CALLER job's `if` is false
the called workflow never starts, so that nested check run is never created —
GitHub records a single run under the bare caller-job name. The poller matched
nothing and burned its full 25-minute timeout, six times over.

`findCheck` now falls back to the bare caller-job name, and only for a completed
`skipped` run. Any other outcome means the caller job ran, so the nested check
run exists and the exact name is still required: this widens the match, not the
pass criteria.

Two problems surfaced while proving it:

- `reusable-await-ts-checks.yml` deliberately checks the poller out from the
  **default branch**, so no pull request can ever exercise a change to it —
  not even one whose base branch carries the fix. Added a `workflow_dispatch`
  route (`check_name` + `pr_head_sha`) that reads the scripts from the
  dispatched ref. It requires write access and touches no PR code, so it adds
  no new trust, and the `pull_request` path still pins the default branch.
  It also gives anyone a way to debug this gate against a real SHA.
- `ts-check-names.test.mjs` ran in **no** workflow. It exists to catch exactly
  this class of bug — a poller naming a check nothing publishes — and it caught
  this branch breaking it, locally, while CI stayed green. Now wired into the
  `security-baseline` suite list (255 tests in that step).

**Proof, dispatched against `7c1edcdcb` — the SHA whose six awaiters timed out:**

| check name | path | run | result |
|---|---|---|---|
| `llm-pr-head-ts-checks / ts-checks` | producer skipped → bare fallback | [35747562981](https://github.com/tetherto/qvac/actions/runs/35747562981) | ✅ pass in ~3s |
| `diffusion-pr-head-ts-checks / ts-checks` | producer skipped → bare fallback | [35747613045](https://github.com/tetherto/qvac/actions/runs/35747613045) | ✅ pass |
| `model-fit-pr-head-ts-checks / ts-checks` | producer ran → exact name | [35747608536](https://github.com/tetherto/qvac/actions/runs/35747608536) | ✅ pass |
| `never-published-pr-head-ts-checks / ts-checks` | nothing publishes it | [35747618312](https://github.com/tetherto/qvac/actions/runs/35747618312) | ✅ **fails closed**, exit 1 after 25 min |

Against the same SHA, `main`'s copy of the script keeps polling past 40s on
`llm-pr-head-ts-checks` (it would hit the 25-minute timeout) and passes
`model-fit-pr-head-ts-checks`, which is the before/after in one pair.

**These awaiters run `main`'s script by design, so the six checks on THIS PR
stay red until it merges.** That is the reason the dispatch route exists.

## 🧪 How was it tested?

Unit and policy suites, plus **28 real AWS Device Farm dispatches** across 12 addons and both platforms. Every result below is read from the run's `test-results.json`, not from the workflow conclusion — each green row is `total=1 passed=1 failed=0` with the named runner recorded `passed`. Each run exercises **one filtered test**, deliberately, to keep Device Farm cost down; this is not full-suite coverage.

- `node --test .github/actions/run-mobile-integration-tests/setup/test/*.test.mjs` — 38 pass (injected HTTP client, no network or token)
- `node --test .github/scripts/test/mobile-prebuild-registry.test.mjs` — 104 pass
- `node --test packages/vla-ggml/scripts/__tests__/*.test.js` — 39 pass
- `node scripts/ci/validate-agent-config.mjs` — 36 skills validated
- `node .github/scripts/lint-workflows.mjs` — actionlint structural gate clean

**Green end-to-end with `prebuild_run_id`**

| addon | platform | run |
|---|---|---|
| llm-llamacpp | Android | [35235330030](https://github.com/tetherto/qvac/actions/runs/35235330030) |
| llm-llamacpp | iOS | [35238417351](https://github.com/tetherto/qvac/actions/runs/35238417351) |
| diffusion-cpp | Android | [35238952152](https://github.com/tetherto/qvac/actions/runs/35238952152) |
| diffusion-cpp | iOS | [35235334565](https://github.com/tetherto/qvac/actions/runs/35235334565) |
| asr-ggml | Android | [35248211731](https://github.com/tetherto/qvac/actions/runs/35248211731) |
| ocr-ggml | Android | [35248215696](https://github.com/tetherto/qvac/actions/runs/35248215696) |
| bci-whispercpp | iOS | [35234285443](https://github.com/tetherto/qvac/actions/runs/35234285443) |
| tts-ggml | iOS | [35234290319](https://github.com/tetherto/qvac/actions/runs/35234290319) |
| translation-nmtcpp | iOS | [35234305241](https://github.com/tetherto/qvac/actions/runs/35234305241) |
| embed-llamacpp | iOS | [35350236485](https://github.com/tetherto/qvac/actions/runs/35350236485) |
| vla *(after the prestage fix)* | Android | [35343789562](https://github.com/tetherto/qvac/actions/runs/35343789562), [35344780875](https://github.com/tetherto/qvac/actions/runs/35344780875) |
| model-fit *(after the fixture fix)* | iOS | [35354826155](https://github.com/tetherto/qvac/actions/runs/35354826155) |

**Existing paths unchanged**

| case | run | evidence |
|---|---|---|
| empty input → published npm | [35234478391](https://github.com/tetherto/qvac/actions/runs/35234478391) (classification), [35235878040](https://github.com/tetherto/qvac/actions/runs/35235878040) (embed) | `Verified: prebuilds come from @qvac/<addon>@<ver> (npm (registry.npmjs.org))` |
| GPR temp-branch pin | [35250764670](https://github.com/tetherto/qvac/actions/runs/35250764670) (model-fit), [35250768920](https://github.com/tetherto/qvac/actions/runs/35250768920) (classification) | `Verified: prebuilds come from @tetherto/<addon>-mono@<ver>-tmp.runid-<id> (pinned, GitHub Packages)` |

**Fail-closed behaviour**

| case | run | result |
|---|---|---|
| run id that built a different addon | [35233951674](https://github.com/tetherto/qvac/actions/runs/35233951674) | `Run … has no 'prebuilds-ocr-ggml' … That run built prebuilds for: vla-ggml.` — `Build mobile app`, `Upload to Device Farm`, `Schedule Device Farm test runs` all **skipped**, zero Device Farm cost |
| both `prebuild_run_id` and `package` | [35250782624](https://github.com/tetherto/qvac/actions/runs/35250782624) | `prebuild_run_id and a pinned package are mutually exclusive (…)` — Device Farm steps skipped |

Also exercised against captured real API payloads for paths the live repo cannot produce: expired artifacts, an unfinished run, a fork-built run, a legacy bare `prebuilds` name, and a run that published nothing.

## 🔐 Action pinning

- `actions/download-artifact`: no version change. The new cross-run download step reuses the SHA already pinned throughout this repo — `3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # 8.0.1` — which is the same pin the `benchmark-perf-*.yml` workflows already use with `run-id`. No third-party action was added, bumped or repinned.

## 🛡️ Permissions changes

- **Scope:** job `build-and-test` in the 12 `integration-mobile-test-*.yml` workflows that expose a pinned-package route.
- **Before:** `contents: read`, `packages: read`, `pull-requests: write`, `id-token: write`
- **After:** the same, plus `actions: read`
- **Justification:** resolving a run and downloading its artifact requires `actions: read`. Without it the lookup 403s *after* the dispatcher has already waited for a build; the resolver detects that case and says so explicitly. No other scope changed, and the token remains `secrets.GITHUB_TOKEN` (not a long-lived PAT). A security review confirmed the run-id path actually *reduces* token exposure on that lane, because it skips the npm/GPR step that materialises a credentialed `.npmrc`.

## 🐛 Bugs found that need a dev (not caused by this PR)

Testing surfaced three failures with isolating evidence. None is introduced by this change, and none blocks it.

**1. `model-fit` — native addon fails to `dlopen` on Android**

[run 35358616440](https://github.com/tetherto/qvac/actions/runs/35358616440) — `logcat_full.txt`, `bare` tag:

```
E bare: Test 'runFitStubTest' failed: AddonError: ADDON_NOT_FOUND:
        Cannot find addon '.' from @qvac/model-fit/binding.js
        Candidates: - linked:libqvac__model-fit.0.12.0.so
        [cause]: Error: dlopen fail
```

Android only — the same test passes on iOS after the fixture fix ([35354826155](https://github.com/tetherto/qvac/actions/runs/35354826155)). Predates this PR: identical failure on a `main` dispatch from before this branch existed ([35212397340](https://github.com/tetherto/qvac/actions/runs/35212397340)).

**2. `embed-llamacpp` — the #4519 binary fails on Android only**

- Android + that binary → fails ([35233800765](https://github.com/tetherto/qvac/actions/runs/35233800765))
- **iOS + the same artifact** → passes ([35350236485](https://github.com/tetherto/qvac/actions/runs/35350236485))
- Android + published `@qvac/embed-llamacpp@0.41.1` → passes ([35235878040](https://github.com/tetherto/qvac/actions/runs/35235878040))

Two variables, each isolated: change the platform or change the binary and it passes. So it is specifically the **Android build** of the run-`35217306279` binary, which comes from PR #4519 ("assert addons pin fabric's C++ runtime and narrow exports") — **merged**, and the published `0.41.1` predates it. Worth checking before embed's next release. Caveat: that run is a *pre-merge* build of #4519, so confirming means a run id from a post-merge build.

**3. `classification-ggml` — dev builds differ from the published release**

- run-id route (`0.26.2` build) → fails ([35248219256](https://github.com/tetherto/qvac/actions/runs/35248219256))
- **GPR route** `@tetherto/classification-ggml-mono@0.26.2-tmp.runid-35247678767` → fails the same way ([35250768920](https://github.com/tetherto/qvac/actions/runs/35250768920))
- published `@qvac/classification-ggml@0.26.1` → passes ([35234478391](https://github.com/tetherto/qvac/actions/runs/35234478391))

It fails through the **pre-existing GPR route** as well as the new one, so this is a real difference in its dev builds rather than anything introduced here.

**4. All Android mobile runs — broken by `bare-type@1.3.0` (filed for @iancris)**

Not caused by this PR, and not addon-specific. The Android app bundle resolves
its `bare-*` native modules from npm unpinned at build time. `bare-type@1.3.0`
was published **2026-09-22 12:07 UTC**; every Android run started after that
links `libbare-type.1.3.0.so` and dies on startup, inside the app bundle's own
`node_modules`, before any qvac addon loads:

```
E bare: ADDON_NOT_FOUND: Cannot find addon '.'
        imported from 'file:///app.bundle/node_modules/bare-type/binding.js'
        Candidates: - linked:libbare-type.1.3.0.so
        [cause]: Error: dlopen failed: cannot locate symbol...
```

Four llm runs, same test-framework commit (`6dfca19f`), four different runners,
two different prebuild bundles. Of the 23 `libbare-*.so` linked into the APK,
`bare-type` is the only one that differs:

| run | date | runner | `libbare-type` | result |
|---|---|---|---|---|
| [35235330030](https://github.com/tetherto/qvac/actions/runs/35235330030) | Sep 17 | runner7 | 1.1.1 | pass |
| [35641773349](https://github.com/tetherto/qvac/actions/runs/35641773349) | Sep 21 | runner9 | 1.1.1 | pass |
| [35735588509](https://github.com/tetherto/qvac/actions/runs/35735588509) | Sep 22 | runner10 | **1.3.0** | fail |
| [35739148483](https://github.com/tetherto/qvac/actions/runs/35739148483) | Sep 22 | runner11 | **1.3.0** | fail |

Controls, both after the publish:

- **llm on `main`**, no pinned package, pre-existing npm fallback — none of this
  PR's code exists there — [35746893107](https://github.com/tetherto/qvac/actions/runs/35746893107): same failure, `qvac__llm` appears **zero** times in logcat.
- **diffusion-cpp** — [35745914476](https://github.com/tetherto/qvac/actions/runs/35745914476): 21 of 23 tests fail with the same signature, `qvac__diffusion` **zero** times. The same addon with the **same** `prebuild_run_id` passed 1/1 on Sep 17 ([35238952152](https://github.com/tetherto/qvac/actions/runs/35238952152)).

Two addons, four runners, both prebuild routes, and a run on `main`. Fix is to
pin the `bare-*` set in the mobile test framework rather than resolve `latest`
at build time.

### Related, worth its own ticket

- **`@qvac/asr-ggml` and `@qvac/audiogen-ggml` publish no mobile prebuilds**, so an empty dispatch input can never work for them. Their `@tetherto/*-mono` dev builds do carry prebuilds, and `prebuild_run_id` works — asr is proven green above.
- **`audiogen-ggml` cannot use `prebuild_run_id` until this merges**, because it loads its composite actions from the default branch. It now fails with an explicit message instead of silently installing `@latest`.
- **The `MODEL_SHARDS` vs `test-groups.json` drift pattern** fixed here for `vla-ggml` could recur in any addon whose prestage generator keeps its own runner list.

## 🔎 Reviews

- **Security review** — no findings at ≥0.8 confidence. Specifically cleared: the run id cannot shape a URL host (digits-only validation before any request), the token never reaches logs, artifacts or the app bundle, and runner names interpolated into the vla `node -e` literal cannot break out (the identifier regex rejects quotes, backslashes, backticks and `$(...)`). Noted that the run-id path *reduces* token exposure, since it skips the step that materialises a credentialed `.npmrc`.
- **Code review** — five findings fixed in `ae2c3d5c9`: newest-artifact tie-break, stale prestage-trap docs, an empty-`RUN_ID` footgun in the documented snippet, the audiogen trim comparison, and the skill catalog entry. One finding not actioned: the 11th `workflow_dispatch` input on the LLM workflow was flagged as exceeding GitHub's 10-input cap, but that cap is no longer enforced — this branch's 11-input workflow dispatched successfully twice ([35235330030](https://github.com/tetherto/qvac/actions/runs/35235330030), [35238417351](https://github.com/tetherto/qvac/actions/runs/35238417351)), and `test-sdk.yml` runs with 25.

### CODE REVIEW

**Review by GustavoA1604 — COMMENTED** [review]
[critical] .github/workflows/integration-mobile-test-ocr-ggml.yml — line 85 (prebuild_run_id:)
                                                                                                                                                      
  > This workflow edit starts the OCR package pipeline, but the TypeScript producer does not create an OCR check for a workflow-only change. The await
  > job treats the resulting skipped/missing check as failure; this PR already produces the same red jobs for OCR, classification, embed, LLM,        
  > translation, and diffusion. Align the TS producer’s path/package mapping with every package workflow trigger, or make the await jobs accept an    
  > intentionally skipped check.                                                                                                                      
                                                                                                                                                      
  [medium] .github/actions/run-mobile-integration-tests/setup/action.yml — line 207 (rm -rf "addon/$ADDON_WORKDIR/prebuilds")                         
                                                                                                                                                      
  > addon-workdir is supplied by the workflows’ manual workdir input without traversal validation. A value such as ../../../otherrepo/otherrepo,      
  > combined with a valid run ID, resolves successfully and recursively deletes a sibling persistent-runner checkout before download-artifact writes  
  > there. Reject absolute paths and .., resolve the target, and verify it remains beneath $GITHUB_WORKSPACE/addon before deleting or downloading.

**Review by sidj-thr — DISMISSED** [review]

File: .github/actions/run-mobile-integration-tests/setup/resolve-prebuild-run.mjs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  const name = segments[segments.length - 1]
+  if (!name || name === '..') return null
+  return `prebuilds-${name}`
+}
+
+// Bare `prebuilds` is the legacy name, kept so an older run id still resolves.
+export function candidateArtifactNames(addonWorkdir) {
+  const merged = mergedArtifactName(addonWorkdir)
+  return merged ? [merged, 'prebuilds'] : ['prebuilds']
+}
```
Comment (iancris) [inline .github/actions/run-mobile-integration-tests/setup/resolve-prebuild-run.mjs]: The run-id prebuild route has no equivalent of the npm route's addon-identity assertion, so a bare `prebuilds` artifact from any addon can be installed under your addon's name and go green.

`candidateArtifactNames` appends the unscoped legacy name for every addon, and `selectArtifact` returns the first candidate holding a **live** row — the expired branch is reached only when no candidate has one. So when `prebuilds-<pkg>` has expired and a bare `prebuilds` row is live, the bare one wins, and the documented "expired fails the run" guarantee holds only while no bare row is live. Nothing downstream re-checks identity: the download selects by id, and *Verify the resolved run's prebuilds cover this platform* only tests that `prebuilds/android-arm64` is non-empty. The npm route guards this with `PACKED_ADDON != EXPECTED_ADDON` (`action.yml:497-504`), whose comment records it "used to exit 0 and put embed's `.bare` on the device".

**Impact:** an older single-addon run id installs that addon's binaries under this addon's name, producing a green Device Farm run over the wrong native binary — the failure class this route exists to close — with a provenance line that truthfully reads `artifact 'prebuilds'`. Dispatch needs repo write and nothing in the tree uploads a bare `prebuilds` today, so it is latent; it stays latent only while the fallback has nothing to serve, which is also the argument for deleting it.

**Suggested fix:** accept the bare name only when `prebuilds-<pkg>` is absent entirely, not merely expired. Checked against the tests — not a drop-in: `test/resolve-prebuild-run.test.mjs:138-148` asserts the opposite for the expired case and `:124-127` pins the candidate list, so both need updating deliberately. The comment at `:146-147` shows the retention tradeoff was weighed; what it does not weigh is that the legacy name carries no addon identity.

Separately, asserting after extraction that the bundle matches the addon's expected `.bare`/`lib<pkg>` stem would mirror the `PACKED_ADDON` check.

File: docs/ci/MOBILE-ON-DEMAND.md [hunk 2]
```diff
@@ -174,12 +260,126 @@ one. Which prebuild depends on the `package` / `package_spec` input:
...
+that run only built the addons it considered affected, and yours was not one. The
+dispatch failure message lists which addons a run did build, so a wrong guess
+tells you where to look.
+
+gh workflow run integration-mobile-test-$WF.yml --repo tetherto/qvac --ref $BRANCH \
+  -f platform=Android \
+  -f devices_custom="Google Pixel 9" \
+  -f device_model_operator=EQUALS \
+  -f prebuild_run_id=$RUN_ID
+```
```
Comment (iancris) [inline docs/ci/MOBILE-ON-DEMAND.md]: The `gh workflow run` example for the PR's flagship new route neither renders as code nor runs as written.

The Route A dispatch command at lines 324-328 has no opening fence. The preceding block closes at 316 and prose runs 318-322, so the command starts unfenced and the bare fence at 329 — meant to close it — opens a block instead. Parity inverts from there: 329→333 wraps the prose line *"The build job's Resolve prebuilds from a run id step prints the provenance:"* in a code block, and 337 opens a block that the ```` ```bash ```` at 388 cannot close, since CommonMark forbids an info string on a closing fence. Lines 338-390 therefore render as one code block: the head-SHA guidance, the "This route fails closed by design" paragraph, the "Three things to know" list, the `#### Route B` heading and Route B Step 1.

Separately, the command references `$WF` and `$BRANCH`, neither defined in Route A — the block above it defines `PKG`, `PR`, `SHA` and `RUN_ID`, while `WF` is first assigned at :402 in Route B and `BRANCH` at :389. Copy-pasting Route A alone dispatches `integration-mobile-test-.yml` with an empty `--ref`.

**Impact:** roughly 53 lines of the document that every `prebuild_run_id` input description points at — including the passage documenting the fail-closed guarantee — render as a code block, and the one command a reader is meant to copy is the only part not in one and does not run.

**Suggested fix:** insert an opening ```` ```bash ```` before line 324. Verified that this alone restores parity for the rest of the file: 329 then closes the command block, 333→337 wraps the provenance sample, and 388→391 wraps the Route B snippet. Then define the two variables locally, matching Route B's block at :401-404 — note `WF` is the workflow slug and cannot be derived from `PKG`, since `integration-mobile-test-vla.yml` builds `packages/vla-ggml`.

**Review by GustavoA1604 — APPROVED** [review]

**Review by iancris — APPROVED** [review]

File: .github/actions/run-mobile-integration-tests/setup/resolve-prebuild-run.mjs (hunk 1 above)
Comment (tobi-legan) [reply .github/actions/run-mobile-integration-tests/setup/resolve-prebuild-run.mjs]: Both fixed.

`selectArtifact` now falls back to the bare `prebuilds` name only when the addon's own bundle is absent from the run entirely — an expired `prebuilds-<pkg>` no longer lets a live bare row win. The two tests you flagged were updated deliberately rather than around.

Also added the post-extraction assertion mirroring `PACKED_ADDON`: it compares the `qvac__<addon>` stem in each platform dir and names the offender.

Checked on hardware, not just in unit tests — dispatched classification-ggml against run 33875701232, whose bare `prebuilds` artifact holds llm-llamacpp:

```
##[error]Run 33875701232's 'prebuilds' contains qvac__llm-llamacpp, not qvac__classification-ggml — it is not a build of this addon.
```

([run 35735689893](https://github.com/tetherto/qvac/actions/runs/35735689893))

File: docs/ci/MOBILE-ON-DEMAND.md (hunk 2 above)
Comment (tobi-legan) [reply docs/ci/MOBILE-ON-DEMAND.md]: Fence parity is fixed in this PR — the opening ```` ```bash ```` before the Route A dispatch is in, and the file is back to 28 balanced fences, so the ~53 lines that were rendering as one code block (including the fail-closed paragraph and the Route B heading) render as prose again.

The undefined `$WF` / `$BRANCH` are a real second bug and I am not claiming them fixed here — that goes out as its own docs PR rather than adding another push to this one. The fix defines both in the Route A block, and unifies the quick start's `ADDON` onto `WF` so the whole document uses one name for the workflow slug, with the `integration-mobile-test-vla.yml` → `packages/vla-ggml` exception called out where it is first used.

Thanks for tracing the parity through line by line — that is what made it a five minute fix.
